src/securejoin/bob.rs

```diff
@@ -110,131 +155,240 @@ pub(super) async fn handle_auth_required(
     context: &Context,
     message: &MimeMessage,
 ) -> Result<HandshakeMessage> {
-    let Some(mut bobstate) = BobState::from_db(&context.sql).await? else {
-        return Ok(HandshakeMessage::Ignore);
-    };
+    // Load all Bob states that expect `vc-auth-required` or `vg-auth-required`.
+    let bob_states: Vec<(i64, QrInvite, ChatId)> = context
+        .sql
+        .query_map(
+            "SELECT id, invite, chat_id FROM bobstate",
+            (),
+            |row| {
+                let row_id: i64 = row.get(0)?;
+                let invite: QrInvite = row.get(1)?;
+                let chat_id: ChatId = row.get(2)?;
+                Ok((row_id, invite, chat_id))
+            },
+            |rows| rows.collect::<Result<Vec<_>, _>>().map_err(Into::into),
+        )
+        .await?;
 
-    match bobstate.handle_auth_required(context, message).await? {
-        Some(BobHandshakeStage::Terminated(why)) => {
-            bobstate.notify_aborted(context, why).await?;
-            Ok(HandshakeMessage::Done)
+    info!(
+        context,
+        "Bob Step 4 - handling {{vc,vg}}-auth-required message."
+    );
+
+    let mut auth_sent = false;
+    for (bobstate_row_id, invite, chat_id) in bob_states {
+        if !encrypted_and_signed(context, message, invite.fingerprint()) {
+            continue;
         }
-        Some(_stage) => {
-            if bobstate.is_join_group() {
+
+        if !verify_sender_by_fingerprint(context, invite.fingerprint(), invite.contact_id()).await?
+        {
+            continue;
+        }
+
+        info!(context, "Fingerprint verified.",);
+        context
```

Comment:
I suggest to reorder sending the handshake message and deleting the bobstate from db which is better e.g. in case of a program crash